cpp/src/io/parquet/writer_impl.cu

```diff
@@ -695,148 +907,49 @@ void writer::impl::write(table_view const &table)
 {
   CUDF_EXPECTS(not closed, "Data has already been flushed to out and closed");
 
-  size_type num_columns = table.num_columns();
-  size_type num_rows    = table.num_rows();
+  size_type num_rows = table.num_rows();
 
-  // Wrapper around cudf columns to attach parquet-specific type info.
-  // Note : I wish we could do this in the begin() function but since the
-  // metadata is optional we would have no way of knowing how many columns
-  // we actually have.
-  std::vector<parquet_column_view> parquet_columns;
-  parquet_columns.reserve(num_columns);  // Avoids unnecessary re-allocation
-
-  // because the repetition type is global (in the sense of, not per-rowgroup or per write_chunk()
-  // call) we cannot know up front if the user is going to end up passing tables with nulls/no nulls
-  // in the multiple write_chunk() case.  so we'll do some special handling.
-  // The user can pass in information about the nullability of a column to be enforced across
-  // write_chunk() calls, in a flattened bool vector. Figure out that per column.
-  auto per_column_nullability =
-    (single_write_mode)
-      ? std::vector<std::vector<bool>>{}
-      : get_per_column_nullability(table, user_metadata_with_nullability.column_nullable);
-
-  uint decimal_precision_idx = 0;
-
-  for (auto it = table.begin(); it < table.end(); ++it) {
-    const auto col        = *it;
-    const auto current_id = parquet_columns.size();
-
-    // if the user is explicitly saying "I am only calling this once", assume the columns in this
-    // one table tell us everything we need to know about their nullability.
-    // Empty nullability means the writer figures out the nullability from the cudf columns.
-    auto const &this_column_nullability =
-      (single_write_mode) ? std::vector<bool>{} : per_column_nullability[current_id];
-
-    parquet_columns.emplace_back(current_id,
-                                 col,
-                                 this_column_nullability,
-                                 user_metadata,
-                                 int96_timestamps,
-                                 decimal_precision,
-                                 decimal_precision_idx,
-                                 stream);
+  auto tbl_meta = (table_meta == nullptr) ? table_input_metadata(table) : *table_meta;
```

Comment:
Since we're making a copy here anyway, it would be better to make a copy when creating the writer - avoid a chance of a dangling pointer. In this case `table_meta` would be a unique pointer (AFAICT).

Reply:
We can create the writer without the table (chunked writer), which is needed for it's construction in case the input was nullptr.

Reply:
I meant that we can copy the input metadata (if any) in the writer constructor, and then in this line we generate it via `table_input_metadata(table)` if it's still null. Basically the only change is that we copy the user metadata sooner.